// cz-git/lib/until/rules.js
"use strict";
/**
 * @description: fork by "@commitlint/cz-commitlint/src/utils/" v16.2.1
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMinLength = exports.getMaxLength = exports.getEnumList = exports.enumRuleIsActive = exports.ruleIsApplicable = exports.ruleIsNotApplicable = exports.ruleIsActive = exports.ruleIsDisabled = void 0;
var share_1 = require("../share");
/**
 * @description: rule is Disabled
 * @example: ruleIsDisabled([0]) => true
 * @example: ruleIsDisabled([2]) => false
 */
function ruleIsDisabled(rule) {
    if (rule && Array.isArray(rule) && rule[0] === share_1.RuleConfigSeverity.Disabled) {
        return true;
    }
    return false;
}
exports.ruleIsDisabled = ruleIsDisabled;
/**
 * @description: rule is use
 * @example: ruleIsActive([0]) => false
 * @example: ruleIsActive([2]) => true
 */
function ruleIsActive(rule) {
    if (rule && Array.isArray(rule)) {
        return rule[0] > share_1.RuleConfigSeverity.Disabled;
    }
    return false;
}
exports.ruleIsActive = ruleIsActive;
/**
 * @description: rule is can ignore
 */
function ruleIsNotApplicable(rule) {
    if (rule && Array.isArray(rule)) {
        return rule[1] === "never";
    }
    return false;
}
exports.ruleIsNotApplicable = ruleIsNotApplicable;
/**
 * @description: rule is effect
 */
function ruleIsApplicable(rule) {
    if (rule && Array.isArray(rule)) {
        return rule[1] === "always";
    }
    return false;
}
exports.ruleIsApplicable = ruleIsApplicable;
function enumRuleIsActive(rule) {
    return (ruleIsActive(rule) && ruleIsApplicable(rule) && Array.isArray(rule[2]) && rule[2].length > 0);
}
exports.enumRuleIsActive = enumRuleIsActive;
function getEnumList(rule) {
    return rule && Array.isArray(rule) && Array.isArray(rule[2]) ? rule[2] : [];
}
exports.getEnumList = getEnumList;
/**
 * @example: getMaxLength(rules['max-header-length'] => 100)
 */
function getMaxLength(rule) {
    if (rule && ruleIsActive(rule) && ruleIsApplicable(rule) && typeof rule[2] === "number") {
        return rule[2];
    }
    return Infinity;
}
exports.getMaxLength = getMaxLength;
/**
 * @example:  getMinLength(rules['min-header-length'] => 2)
 */
function getMinLength(rule) {
    if (rule && ruleIsActive(rule) && ruleIsApplicable(rule) && typeof rule[2] === "number") {
        return rule[2];
    }
    return 0;
}
exports.getMinLength = getMinLength;
